### ISSUE TITLE
[Proposal] Replace React Class (& Maybe Also JS OOP) with Functional

### ISSUE THREAD

**Issue by vegetabill** [Techtonica/curriculum#1550]
## TL;DR

- JS OOP is unpopular and awkward. We should just skip it.
- React Functional is the ~future~ present. We shouldn't teach class components.

In my opinion, these both are tightly related and we should both or neither, but that's of course debatable 😄 .

This idea has been pitched several times and every new SEAM (including me) expresses something along these lines, and it's a good idea. _But_ it is also a combination of intentional curriculum principles and logistics that complicate this. (See Pros & Cons below)

This specific latest proposal comes from a Slack thread by @gsong :

> Hi everyone, @SarahCawley and I were chatting last night and it seems like the current curriculum is teaching React class-based components. While it’s too late to unwind the curriculum, it’s not too late to unwind the teaching and move the cohort to functional components and hooks.
> They are going to run into a lot of problems learning class-based components, as all of the newer examples and libraries they’ll run into will be functional and more importantly, hook based.
> :exclamation: Class-based components cannot play in the hooks ecosystem.
> One thought is for me to lead a one hour workshop/study group each week, but that’s not nearly enough. We’ll need more people to reinforce teaching functional components and hooks.
> Thoughts? Ideas?
> A side corollary (lower priority): I think we should shy away from object-oriented patterns where it doesn’t make sense. For example, the pattern feels awkward in the Eventonica project from what I’ve seen

## Pros
1. First-time language learners benefit from sticking to one paradigm. Since we're teaching JS, if you're picking one it would 100% be functional.
1. React class components do not interop with new versions of libraries. Hooks are where everything is going.
1. Eventually we will have to do this, and deferring it doesn't necessarily make the task smaller.
1. (For general JS OOP) Learning OOP is helpful when doing interview questions that use OOP (often data structures) and if the participants land a Java or other OO-language internship.

## Cons
1. The original curriculum, due to when it was written, focused on class-based components but also threw in the then-new functional feature (no hooks yet). Although interchangeable to an experienced engineer, this essentially doubled the learning (and error) surface area. **This is the main reason we consolidated to mainly focus on class-based components** with just a hooks/functional elective later.
1. Hooks remain a little magical even to experienced engineers. The class-based components fit  into nascent mental models more easily, (which might offset that they are dated, i.e. better to master an old form than use a newer form without understanding it as well)
1. Real codebases the participants will encounter in their internships will almost certainly have some class-based component code

A con that is not specific to this idea, but worth noting, is that it has been historically challenging to make large scale changes to the curriculum, both in the sheer person-hours involved and the inevitable inconsistencies it creates if you don't make a sweeping change.

**Comment by alodahl** [comment]
I'm on board as well.  I think class-based react components can be mentioned, and participants should be familiar with how it looks, but can be spared having to build with it or learn it deeply.  I was thinking of keeping the existing examples, setting them aside in a different file, and referring to them every once in a while, "and this is how it would look in a class-based react component".

**Comment by vegetabill** [comment]
I believe this is happening so we don't need this discussion ticket anymore.
